packages/ui-kit/lib/components/pre-assessment/form-renderer.tsx

```diff
@@ -0,0 +1,193 @@
+                    value = (singleChoice?.options || []).map(opt => ({
+                        name: opt.name,
+                        isSelected: opt.isSelected
+                    }));
+                    break;
+                }
+                case FormElementType.RichText: {
+                    const richText = formElement as RichTextElement;
+                    value = richText?.content || '';
+                    break;
+                }
+                default:
+                    value = '';
+            }
+
+            const isValid = formElements[element.type].validate(element, value);
+
+            if (!isValid) {
+                newErrors[element.id] = true;
+            }
+        }
+
+        // Update errors state
+        setFormErrors(newErrors);
+
+        // Return true if no errors were found
+        return Object.keys(newErrors).length === 0;
+    };
+
+    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
+        if (e) e.preventDefault();
+        e.stopPropagation();
+        // Validate the form before submission
+      
+        if (!validateForm()) {
+            // If validation fails, don't submit
+            return;
+        }
+
+        onSubmit(formValues.current);
+    };
+
+    const getErrorMessage = (element: FormElement): string => {
+        switch (element.type) {
+            case FormElementType.TextInput:
+                return dictionary.components.formRenderer.fieldRequired;
+            case FormElementType.SingleChoice:
+                return dictionary.components.formRenderer.selectOption;
+            default:
+                return dictionary.components.formRenderer.fieldRequired;
+        }
+    };
+
+    return (
+        <form onSubmit={handleSubmit} className="max-w-[560px] p-6 flex flex-col gap-4 bg-card-fill shadow-[0px_4px_12px_0px_base-neutral-950] border-1 rounded-medium border-card-stroke text-text-primary">
+            <h3>{dictionary.components.formRenderer.title}</h3>
+            <div className="flex flex-col gap-4">
+                {elements.map((elementInstance) => {
+                    const Element = formElements[elementInstance.type].formComponent;
+                    return (
+                        <div key={elementInstance.id.toString()} className="flex flex-col gap-2">
+                            <Element
+                                submitValue={submitValue}
+                                elementInstance={elementInstance}
+                                locale={locale}
+                            />
+                            {formErrors[elementInstance.id] && (
+                                <Banner
+                                    style="error"
+                                    description={getErrorMessage(elementInstance)}
+                                />
+                            )}
+                        </div>
+                    );
+                })}
```

Comment:
Just to note - we don't sort elements by their order here. I am fine with doing this outside the form renderer, but we should keep it in mind